### DISCUSSION TITLE
Support calling AWS generate-db-auth-token to get DB password for AWS RDS/Aurora Postgres

### DISCUSSION THREAD

**Discussion by krestivo-kdinfotech** [grafana/grafana#48170] [Data Sources]
**What would you like to be added**:

Support calling AWS generate-db-auth-token API to obtain a Postgres password before opening or re-opening any DB connection.

**Why is this needed**:
When running in AWS with IAM auth, you don' t have a password, you use a temporary token which goes into the PGPASSWORD field of the Postgres DB connection. This token times out after 15 minutes.

Grafana seems to close and open DB connections often, even with pooling, rather than keeping a single connection open. When this connection goes stale, Grafana seems to attempt to reconnect to the database. However, the token is no longer any good and you get RDS password errors.

The fix would be for Grafana to support a hook that could be run prior to it attempting to connect to a database or reconnect to a database, which the user could specify an external (shell) function to obtain the correct RDS token.

**Comment by krestivo-kdinfotech** [comment]
I would be happy to add this feature if someone would like to point me in the general direction of where in the codebase it *should* go.

**Comment by marefr** [comment]
Find it very unlikely that we'll add your suggested solution to Grafana. Haven't received any other reports regarding this problem which I find weird. Maybe there's another solution to your problem like using a sql proxy or similar? Will close this for now, but we'll consider reopen this should it get more traction/activity/upvotes.

**Comment by samwise-nl** [comment]
I would request that this be looked into as well. If IAM auth was supported we'd be implementing Grafana right now instead of trying to find a solution that supports it properly.

**Comment by krestivo-kdinfotech** [comment]
SQLproxy would likely work but it is a hack (and another moving part). We ended up coming up with a simpler hack: a magic static password that we now have to manage.

**Comment by kinghuang** [comment]
Can this be reconsidered? I have RDS Aurora databases with IAM authentication, and would appreciate Grafana calling `generate-db-auth-token` instead of generating a static password.

**Comment by theis188** [comment]
Adding my vote as well. This would be a useful feature. The lack of this feature is causing me to look into alternate dashboarding solutions due to policies against static passwords for rds.

**Comment by apolegoshko** [comment]
I was able to connect pod running in EKS with attaching IAM Role, that have rights to access AWS RDS Postgresql instance using token. Of course it expired after 15 min. This feature will be very useful in the future.

**Comment by alexjfisher** [comment]
> Haven't received any other reports regarding this problem which I find weird.

@marefr There was https://community.grafana.com/t/how-do-i-setup-iam-authentication-for-a-backing-rds-mysql-database/5767 and https://stackoverflow.com/questions/48996897/how-do-i-setup-grafana-on-an-ec2-instance-using-iam-auth-for-an-rds-mysql-instan so I think quite a few people have tried to get this to work and would appreciate this ticket being reopened?

**Comment by v1ctorrhs** [comment]
Adding my vote here. We try to stay away from storing passwords and rely on IAM auth for most of our services.

**Comment by bishwashrestha** [comment]
adding my votes well (similar use cases as stated by other commenters) - let's reopen this please?

**Comment by gabrielsyapse** [comment]
please reopen

**Comment by rayshard-thompson** [comment]
Adding my vote as well. Would be nice if this was reopened.

**Comment by chaosun-abnormalsecurity** [comment]
Please consider to add this support

**Reply by keymon** [reply]
:+1:  Plus one here, we need this to comply with the security requirements.

**Comment by m-ghazanfar** [comment]
Please consider adding this or any other way to support rotating credentials. Need it to comply with security requirements.

**Comment by lily-es** [comment]
Would be really nice if this was supported. Please consider adding this feature.

**Comment by danielthedm** [comment]
Moved to issue: https://github.com/grafana/grafana/issues/75965
